### PR TITLE
Optionally Use Search Engine from JEI or REI

### PR DESCRIPTION
**Pull request by MarkusTieger**
Allows to use the Search Engine from JEI or REI.

This can be useful because it is annoying when you try to search something with the modid and the name like it is possible in JEI. For example "@ae2 crystal". Without it, it tries to search a mod with the name "ae2 crystal".
It is configurable in the Terminal Settings. (opt-out to disable btw)

### CODE REVIEW

File: src/main/java/appeng/client/gui/me/search/SearchPredicates.java [hunk 1]
```diff
@@ -1,17 +1,110 @@
...
+import mezz.jei.api.runtime.IJeiRuntime;
+import net.minecraft.core.Registry;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
 
 final class SearchPredicates {
 
+    static Predicate<GridInventoryEntry> fromStringUsingJEI(String searchString, IJEI jei) {
```
[the remark this answers is not in the document]
Comment (MarkusTieger) [reply src/main/java/appeng/client/gui/me/search/SearchPredicates.java:40]: IJEI is an ae2 class btw

File: src/main/java/appeng/client/gui/me/search/SearchPredicates.java [hunk 2]
```diff
@@ -1,17 +1,98 @@
...
+        IJeiRuntime runtime = ((JeiRuntimeAdapter) jei).getRuntime();
+
+        var filter = runtime.getIngredientFilter();
+
+        var old = filter.getFilterText();
+        filter.setFilterText(searchString);
+
+        var filtered = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (MarkusTieger) [reply src/main/java/appeng/client/gui/me/search/SearchPredicates.java:35]: I don't see why it's needed for type "Object"

File: src/main/java/appeng/client/gui/me/search/SearchPredicates.java [hunk 3]
```diff
@@ -1,17 +1,98 @@
...
+
+            filtered.addAll(
+                    filter.getFilteredIngredients(type)
+                            .stream()
+                            .map((e) -> subtype.getBase(e))
+                            .toList());
+        }
+
+        if (!AEConfig.instance().isSyncWithExternalSearch())
+            filter.setFilterText(old);
```
[the remark this answers is not in the document]
Comment (MarkusTieger) [reply src/main/java/appeng/client/gui/me/search/SearchPredicates.java:49]: Else the current search in jei will be overriden with the ae2 search, even if "sync with external search" is off. Thats the "fix" for it.

**Comment by MarkusTieger** [comment]
> @MarkusTieger Sorry for not coming back to you earlier. I am in the process of moving master to Neoforge, so it'll take a while before I can get to this PR. I also would like to actually refactor the functionality significantly, namely that "use JEI search" (the BIG radio buttons we currently have) should probably use this functionality instead of just hiding our own search box, but then use different filter logic.
> 
> All in all: it will take a while 😓

no problem

**Comment by Relvl** [comment]
Any progress? We all very need this change! :)

**Comment by MarkusTieger** [comment]
> Any progress? We all very need this change! :)

I almost forgot this issue, thanks for reminding me. This week I am busy, but I should be able to update this pr in the next week.

**Comment by Relvl** [comment]
Also maybe add JEI sorting function? #7827

**Comment by MarkusTieger** [comment]
I guess then this issue is no longer required and https://github.com/AppliedEnergistics/Applied-Energistics-2/pull/7883 will resolve the issue with the bad search?

**Pull request closed without merging**